src/index/CompressedRelation.cpp

```diff
@@ -458,51 +458,64 @@ auto getRelevantIdFromTriple(
+auto CompressedRelationReader::getBlocksForJoin(
     ql::span<const Id> joinColumn,
-    const ScanSpecAndBlocksAndBounds& metadataAndBlocks) {
-  // We need symmetric comparisons between Ids and blocks.
+    const ScanSpecAndBlocksAndBounds& metadataAndBlocks)
+    -> GetBlocksForJoinResult {
+  if (joinColumn.empty() || metadataAndBlocks.getBlockMetadataView().empty()) {
+    return {};
+  }
+
+  // `id < block` iff `id < block.firstTriple`
   auto idLessThanBlock = [&metadataAndBlocks](
                              Id id, const CompressedBlockMetadata& block) {
     return id < getRelevantIdFromTriple(block.firstTriple_, metadataAndBlocks);
   };
 
+  // `block < id` iff `block.lastTriple < id`
   auto blockLessThanId = [&metadataAndBlocks](
                              const CompressedBlockMetadata& block, Id id) {
     return getRelevantIdFromTriple(block.lastTriple_, metadataAndBlocks) < id;
   };
 
-  auto lessThan = ad_utility::OverloadCallOperator{
-      idLessThanBlock, blockLessThanId, BlockLessThanBlock{}, std::less<Id>{}};
-
-  // Find the matching blocks by performing binary search on the `joinColumn`.
-  // Note that it is tempting to reuse the `zipperJoinWithUndef` routine, but
-  // this doesn't work because the implicit equality defined by
-  // `!lessThan(a,b) && !lessThan(b, a)` is not transitive.
-  auto blockIsNeeded = [&joinColumn, &lessThan](const auto& block) {
-    return !ql::ranges::equal_range(joinColumn, block, lessThan).empty();
-  };
-
   std::vector<CompressedBlockMetadata> result;
-  ql::ranges::copy(metadataAndBlocks.getBlockMetadataView() |
-                       ql::views::filter(blockIsNeeded),
-                   std::back_inserter(result));
+  const auto& mdView = metadataAndBlocks.getBlockMetadataView();
 
-  // The following check is cheap as there are only few blocks.
-  AD_CORRECTNESS_CHECK(std::unique(result.begin(), result.end()) ==
-                       result.end());
-  return result;
+  auto [colIt, colEnd] = getBeginAndEnd(joinColumn);
+  auto [blockIt, blockEnd] = getBeginAndEnd(mdView);
+  GetBlocksForJoinResult res;
+
+  // Manually count the blocks that we pass in the `mdView`.
+  auto& blockIdx = res.numHandledBlocks;
+  while (true) {
+    while (colIt != colEnd && idLessThanBlock(*colIt, *blockIt)) {
+      ++colIt;
+    }
+    if (colIt == colEnd) {
+      return res;
+    }
+    while (blockIt != blockEnd && blockLessThanId(*blockIt, *colIt)) {
+      ++blockIt;
+      ++blockIdx;
+    }
+    if (blockIt == blockEnd) {
+      return res;
+    }
+    // It holds that `*blockIt >= *colIt`. As the entries in the `joinColumn`
+    // are sorted, it suffices to additionally find the values where
+    // `*blockIt <= *colIt` to find possibly matching blocks.
+    while (blockIt != blockEnd && !idLessThanBlock(*colIt, *blockIt)) {
+      res.matchingBlocks_.push_back(*blockIt);
+      ++blockIt;
+      ++blockIdx;
+    }
+    if (blockIt == blockEnd) {
```

Comment:
Unimportant for this PR, but I wonder if there's a more aesthetic way to represent this while pattern.

Reply:
Some of the `while` loops can be `it = std::find_if(it, end, lambda)` which has exactly the same number of lines. 
the
`if (isEnd) { return}` could only be shortened by a macro. 
So I will leave this for now as is.
With a generic "merge" algorithm some things could be done, but then again the counting of the passed blocks has to be customized etc...